Fix the call node max depth calculation

gregtatum opened the pull request:
I have a few yak shaving PRs for my dhat importer in #3128. This is the first, which fixes some issues around the max depth calculation.

Notice the extra white space on the flame graph when moving the viewport to the top: https://share.firefox.dev/3qNCQ09
This is fixed when taking into account the preview filtered thread.

I went through all of the uses of the max depth, and ensured they were correct.

I'm flagging both Julien and Nazim for review, as I don't know your workloads right now, but I only need one person to look at it.

gregtatum commented on src/profile-logic/profile-data.js:
```diff
@@ -1787,18 +1787,25 @@ export function convertStackToCallNodePath(
  */
 export function computeCallNodeMaxDepth(
-  thread: Thread,
+  samples: SamplesLikeTable,
```
This is also a fix for the dhat PR. The call node max depth needs the current samples-like table, NOT the thread to compute this correctly.

gregtatum commented on src/selectors/per-thread/stack-sample.js:
```diff
@@ -214,31 +208,50 @@ export function getStackAndSampleSelectorsPerThread(
+    CallTree.extractSamplesLikeTable
+  );
+
+  const getPreviewFilteredSamplesForCallTree: Selector<SamplesLikeTable> = createSelector(
```
I created these with explicit names, as I found myself confused on what was used where.

gregtatum commented on src/test/store/profile-view.test.js:
```diff
@@ -1828,9 +1828,18 @@ describe('snapshots of selectors/profile', function() {
+    const { getState } = setupStore();
+    expect(
+      selectedThreadSelectors.getPreviewFilteredCallNodeMaxDepth(getState())
+    ).toEqual(4);
```
This is my end of the day, and I didn't really have time to write better tests than this.

gregtatum replied:
I added a better EmptyReasons tests.

julienw commented:
Thanks for the patch @gregtatum !

I don't see issues with the code, but I have a few comments and questions before moving forward.

First: can you please precise why this is useful? Especially is it blocking the rest of the work? This question is to balance with the first issue I'm seeing (see below).

Second: I see this is breaking one case, which may be unimportant, but I'd like your feedback:

1. [open the flame graph on this profile](https://deploy-preview-3221--perf-html.netlify.app/public/8tzsbbb2mtvdebtad4r760re887mg5f3wn0md08/flame-graph/?globalTrackOrder=12-0-1-2-3-4-5-6-7-8-9-10-11&hiddenGlobalTracks=12-0-1-2-3-4-5-6-7-8-9-10&hiddenLocalTracksByPid=4453-0-1-2&localTrackOrderByPid=4404-2-3-0-1-4-5~15157-0-1~4426-0-1~15136-0-1~4452-0-1~4449-0-1~4448-0-1~14900-0-1~4450-0-1~4427-0-1~4892-0-1-2~4453-0-1-2~&profileName=cnn.com%20Sep%202020&thread=13&transforms=df-37&v=5)
2. move up in the flame graph
3. make a preview selection while looking at the flame graph

=> This is resetting at the bottom (presumably because it tries to compute the max depth first with very few samples?). But if you try [with the production branch](https://profiler.firefox.com/public/8tzsbbb2mtvdebtad4r760re887mg5f3wn0md08/flame-graph/?globalTrackOrder=12-0-1-2-3-4-5-6-7-8-9-10-11&hiddenGlobalTracks=12-0-1-2-3-4-5-6-7-8-9-10&hiddenLocalTracksByPid=4453-0-1-2&localTrackOrderByPid=4404-2-3-0-1-4-5~15157-0-1~4426-0-1~15136-0-1~4452-0-1~4449-0-1~4448-0-1~14900-0-1~4450-0-1~4427-0-1~4892-0-1-2~4453-0-1-2~&profileName=cnn.com%20Sep%202020&thread=13&transforms=df-37&v=5) this stays stable.

I know that using the preview selection while looking at the flame graph is a way to analyze that you were using a lot for memory allocations. But I don't know if moving up the flamegraph is a common enough use case to be important here.

Lastly: I was a bit afraid for the performance (looking at all samples for each preview, vs looking all call nodes only once), but a rough visual testing on a busy profile showed no difference (I didn't profile though). So I'm not really worried about that anymore. (for reference, here are the links for the busy profile: [deploy preview](https://deploy-preview-3221--perf-html.netlify.app/public/2b38186f16b0ab99e3c691b2b04869543b40e504/flame-graph/?globalTrackOrder=0-1&localTrackOrderByPid=4176-1-2-0~6176-0-1~&thread=2&v=5) / [production](https://profiler.firefox.com/public/2b38186f16b0ab99e3c691b2b04869543b40e504/flame-graph/?globalTrackOrder=0-1&localTrackOrderByPid=4176-1-2-0~6176-0-1~&thread=2&v=5))

Happy to get your feedback on all this!

Thanks again

gregtatum commented:
* [Before](https://profiler.firefox.com/public/8tzsbbb2mtvdebtad4r760re887mg5f3wn0md08/flame-graph/?globalTrackOrder=12-0-1-2-3-4-5-6-7-8-9-10-11&hiddenGlobalTracks=12-0-1-2-3-4-5-6-7-8-9-10&hiddenLocalTracksByPid=4453-0-1-2&localTrackOrderByPid=4404-2-3-0-1-4-5~15157-0-1~4426-0-1~15136-0-1~4452-0-1~4449-0-1~4448-0-1~14900-0-1~4450-0-1~4427-0-1~4892-0-1-2~4453-0-1-2~&profileName=cnn.com%20Sep%202020&range=1610m1711~2227m97&thread=13&transforms=df-37&v=5)
 * [After](https://deploy-preview-3221--perf-html.netlify.app/public/8tzsbbb2mtvdebtad4r760re887mg5f3wn0md08/flame-graph/?globalTrackOrder=12-0-1-2-3-4-5-6-7-8-9-10-11&hiddenGlobalTracks=12-0-1-2-3-4-5-6-7-8-9-10&hiddenLocalTracksByPid=4453-0-1-2&localTrackOrderByPid=4404-2-3-0-1-4-5~15157-0-1~4426-0-1~15136-0-1~4452-0-1~4449-0-1~4448-0-1~14900-0-1~4450-0-1~4427-0-1~4892-0-1-2~4453-0-1-2~&profileName=cnn.com%20Sep%202020&range=1610m1711~2227m97&thread=13&transforms=df-37&v=5)

Correctness:
* Before: The committed range selection displays too much whitespace above due to incorrect max call node depth calculation.
* After: There is no additional whitespace above the graph. The graph only goes to the top.

Fixed:
* Create a preview selection with the "After". The flame graph uses the filtered call node max depth, which makes the viewport stable.

Empty reasons:
* Before: Create a preview selection where there are no samples. The view is blank with no reason.
* After: Create a preview selection where there are no samples. The view shows empty reasons

gregtatum commented on src/components/calltree/CallTree.js:
```diff
@@ -334,7 +334,12 @@ export const CallTree = explicitConnect<{||}, StateProps, DispatchProps>({
     invertCallstack: getInvertCallstack(state),
     implementationFilter: getImplementationFilter(state),
-    callNodeMaxDepth: selectedThreadSelectors.getCallNodeMaxDepth(state),
+    // Use the filtered call node max depth, rather than the preview filtered call node
```
I added justifications for all of my choices here.

gregtatum commented on src/selectors/per-thread/stack-sample.js:
```diff
@@ -84,12 +84,6 @@ export function getStackAndSampleSelectorsPerThread(
   );
 
-  const getCallNodeMaxDepth: Selector<number> = createSelector(
-    threadSelectors.getFilteredThread,
```
Here's the REAL bug before I did all this other work. This was passing in the thread, NOT the samples-like table. When viewing a native allocations profile, this gives the very wrong result. As long as the stacks in the samples are as deep as the native allocations, this is fine. However, in my dhat importer, the samples have a stack depth of 0, while the native allocations have a large stack depth.

gregtatum replied:
However, the other changes I made were related to correctness and being explicit about what is being measured here. I left a few STR in my main comment outlining things that were subtly incorrect.

gregtatum commented:
> Second: I see this is breaking one case, which may be unimportant, but I'd like your feedback:

Yes, this was bad. Thanks for catching it! I moved the computation to the filtered thread max depth. This means preview selections can have extra whitespace, but are limited by the committed selection's max depth. This feels like the correct solution here. I also switched the call tree to use the filtered thread max depth. I don't think this one matters as much, but it'll help the call tree from changing width. This seems less important.

> Lastly: performance.

I agree it's a bit unfortunate to have to walk the samples, but I think it's the correct thing to do here to get this value. The committed range selection is pretty stable, so I don't think that'll be too much of a perf hit. The flame graph empty reason is the only remaining consumer of the preview selection's max depth.

julienw commented:
> > Second: I see this is breaking one case, which may be unimportant, but I'd like your feedback:
> 
> Yes, this was bad. Thanks for catching it! I moved the computation to the filtered thread max depth. This means preview selections can have extra whitespace, but are limited by the committed selection's max depth. This feels like the correct solution here. I also switched the call tree to use the filtered thread max depth. I don't think this one matters as much, but it'll help the call tree from changing width. This seems less important.

Thanks!

> 
> > Lastly: performance.
> 
> I agree it's a bit unfortunate to have to walk the samples, but I think it's the correct thing to do here to get this value. The committed range selection is pretty stable, so I don't think that'll be too much of a perf hit. The flame graph empty reason is the only remaining consumer of the preview selection's max depth.

Yeah, I'm not too concerned now that we use the committed range.
Also I was today years old when I realized that the CallNodeInfo doesn't change when filtering samples (but it does with anything changing the stackTable like transforms / inversion)... but we still regenerate it. I wonder if we should memoize it (`createSelector` memoizes only using the `thread` object which do change in all these cases).

julienw reviewed (approved):
Thanks, this looks good to me!

gregtatum commented:
> Also I was today years old when I realized that the CallNodeInfo doesn't change when filtering samples

Yeah, this was intentional from me and Markus, as it's only invalidated when the actual stack structure changes. I had forgotten when working on this though!
